batchdelete: allow deletion of subpages

siddharthvp opened the pull request:
Trying to solve #305. Though I can't be sure whether this works without having the sysop bit.

This handles the subpages just as the redirects are being handled (they are deleted without being listed, so individual pages can't be de-selected). We may need to think about whether sysops ought to be deleting subpages without knowing anything about which are the subpages or how many of them are there (until they see the "Deleting ___" status messages). If deemed necessary, this can be tweaked to list out all the subpages along with the main pages (so that they are known to the user beforehand, and can be de-selected).

Amorymeltzer commented:
Haha great minds: I'm on a bus with limited wifi, and figured this was an easy thing to whip up while watching Firefly.  This is nearly exactly what I wrote up — I've got a couple longer comments I'll make when I get a chance, but overall structure is fine.

JJMC89 commented:
This shouldn't be done in namespaces where subpages are not enabled, like ns:0 on enwiki.

I think it would be better to have a selection like d-batch.

Example case:
- Article to be deleted: `Foo`
- Another article at `Foo/bar` should not be deleted.
- `Talk:Foo` and `Talk:Foo/Archive n` should be deleted but not `Talk:Foo/bar` and `Talk:Foo/bar/Archive n`.

Similar things occur with templates where a template name contains a `/` but isn't part of and shouldn't be deleted when the root template is.

Amorymeltzer commented:
That basically covers what I was going to say, namely this ''needs'' to ignore Mainspace.  I don't quite follow on templatespace, though — @JJMC89 do you have an example?

Anyway, the other major difference between this and my code was I tack on a `delete_talkpages` to each subpage; that is, if we're deleting subpages of `User:DouglasAdams`, then `User:DouglasAdams/42` as well as `User talk:DouglasAdams/42` should be deleted.

That's straightforward enough, but the tricky issue is subpages of talk pages without corresponding pages, aka archives.  I basically dealt with this by *not* dealing with it.  I think we should include some notes/tooltips, but basically they're currently not being deleted anyway so it's no loss.  Moreover, I don't know that we want to automatically remove all talk page archives, and I think it'd be better to have sysops have to explicitly note the existence of talk pages in their list.  How's that sound @JJMC89?

JJMC89 commented:
> I don't quite follow on templatespace, though — @JJMC89 do you have an example?

I couldn't find the original example that I was thinking of, but [Template:Yes/no](https://en.wikipedia.org/w/index.php?title=Template:Yes/no&redirect=no) (redirect to Template:Yesno) isn't related to [Template:Yes](https://en.wikipedia.org/wiki/Template:Yes).

> That's straightforward enough, but the tricky issue is subpages of talk pages without corresponding pages, aka archives. I basically dealt with this by _not_ dealing with it. I think we should include some notes/tooltips, but basically they're currently not being deleted anyway so it's no loss. Moreover, I don't know that we want to automatically remove all talk page archives, and I think it'd be better to have sysops have to explicitly note the existence of talk pages in their list. How's that sound @JJMC89?

I'm not really too concerned with talk archives. In most cases, pages being deleted don't have them.

> I think the three obvious options are:
> 
> * Only enable subpages for certain namespaces (easy)
> * Require manual confirmation for every list of subpages for every page (medium)
> * Dump them all in a new dbatch list and delete _that_ list (hard)

I don't think deleting all subpages without confirming what is being deleted is a good idea, and I would prefer to be able to select which subpages to delete.

siddharthvp commented:
> Dump them all in a new dbatch list and delete that list (hard)

I think this option is the way to go. I spent a couple of hours on this and hope to fix it up soon. I don't think there is a way to list out all the subpages of n different pages using a single API call; it requires n API calls. So @atlight's `Morebits.batchOperation` needs to be used. The trouble with the batchOperation framework is that it does not give an option to provide a callback ( which we need to list the pages on the interface) that executes after the whole batch has processed. I guess I'll have to add that ability to the batchOperation class.

Amorymeltzer commented:
@JJMC89 I see now — the problem being, of course, that the software thinks [[Template:Yes/no]] *is* a subpage of [[Template:Yes]].  I think templatespace would be one of the man places this would be desirable, with so many /doc and other subpages frequently being left behind.  Perhaps we'll have to rely upon sysops to be careful?  I'm in favor of loading this up with a couple scary warnings.

siddharthvp commented:
-

siddharthvp commented:
@Amorymeltzer @JJMC89 Done. The list of all subpages will now be displayed below the main list. 

Archives of talk pages won't be deleted. If we wish to delete them too, we'd have to query the list of subpages for all talk pages too, more than doubling the number of api calls. 

I've provided additional options for i) deletion of redirects of subpages, and ii) unlinking of subpages, but not for deletion of talk pages of subpages (which is done if main talk page deletion is checked).

Amorymeltzer commented:
I'm heading to bed so can't do a full review, but this will display subpages of every page, regardless of whether they're checked or not, right?  That seems problematic.  I don't think we want to regenerate the subpage list on the fly whenever one of the top-level page boxes is checked or unchecked, but the list of subpages *should* respect that list somehow.  Open to brainstorming.

siddharthvp commented:
> but this will display subpages of every page, regardless of whether they're checked or not, right?

Should have thought about that earlier. Fixed now.

Amorymeltzer commented:
@siddharthvp I don't have the time/brainpower atm to delve into the code, but from a quick look and test, there's something weird happening when checking/unchecking boxes.

[Test page](https://test.wikipedia.org/wiki/User:Amorymeltzer/sandbox) has two pages linked, one mainspace and one user, the latter with subpages.

1.  Clicking dbatch opens the Twinkle menu with both checked (good)
2.  Clicking subpages loads subpages, all checked (good); unchecking subpages returns just the two pages, both checked (good).  Toggling back and forth, they all remain checked (good)
3.  Unchecking or checking either page (either before or after toggling subpages) and then toggling subpages, they each unchecked (good)
4.  Subsequent re-checking of either page, however, is lost when subpages is toggled (bad)
5.  The above resets if the window is closed and reopened.

If I toggle a page on or off before toggling subpages, there's no issue, but it seems toggling the subpage box will retain a page's unchecked status if reached once the subpages menu is triggered.

siddharthvp commented:
Did you use the Select All/Deselect All buttons any time during the above tests? I belatedly realised that those buttons don't work properly in the presence of subpage checkboxes. I will fix that. 

I also plan to add arrow `>` links next to the subpages (as we do in Tag module) for easy navigation to those pages. I too am very busy over the next 2 weeks, as you may have judged by my lack of response at the PRs where you've pinged me recently. I'll deal with this and the others whenever I get time.

Amorymeltzer commented:
No, just the checkboxes.  And no rush, thanks for letting me know.

siddharthvp commented on modules/twinklebatchdelete.js:
```diff
@@ -59,7 +61,25 @@ Twinkle.batchdelete.callback = function twinklebatchdeleteCallback() {
+											label: 'Delete redirects to deleted subpages',
+											name: 'delete_subpage_redirects',
+											value: 'delete_subpage_redirects',
+											checked: true
```
Nice idea!

siddharthvp commented:
I've just pushed a new version, which is hopefully bug-free. Fixed the issue that you pointed out in bullet point 4 above. Fixed the select all/deselect all buttons. Fixed the issue with shiftclick support. Added `>` links to pages.

siddharthvp commented:
In order to fix the issue of shiftclick support to the main page list (when subpages are present), I have provided a fix in `Morebits.checkboxShiftClickSupport`, so that is the 3rd change to morebits in this PR.  Basically, the issue was that when you try to uncheck multiple pages at once, though those pages do get unchecked, their lists of subpages didn't get hidden. (This was because `checkbox.checked = true/false` does not trigger the `click` event, whereas `checkbox.click()` does.

siddharthvp commented:
In the present scheme, the subpages are being deleted only after all the main pages (along with main pages' talkpages and redirects) have been deleted. It'd be better to delete subpages along with their parent page. But there seems to be no easy way of identifying which subpages belong to which page at evaulate-time.

siddharthvp commented on modules/twinklebatchdelete.js:
```diff
@@ -352,6 +544,22 @@ Twinkle.batchdelete.callbacks = {
 		page.setEditSummary("[[WP:CSD#G8|G8]]: [[Help:Talk page|Talk page]] of deleted page \"" + apiobj.params.page + "\"" + Twinkle.getPref('deletionSummaryAd'));
 		page.deletePage();
 	},
+	deleteSubpages: function( apiobj ) {
```
no longer being used. This was a remnant from the early version where we weren't listing subpages at all.

siddharthvp commented:
note: I deleted the commit that added `HTMLFormElement.prototype.getUnchecked`, as it has already been added to morebits by the now-merged #485.

siddharthvp commented:
Have now tested this using the newly-gained admin bit on testwiki. Everything seems to work as intended.

Amorymeltzer commented:
My only other real nitpick is that if someone opens the dbatch menu, loads the subpage box, closes the dbatch menu, then re-opens it, the whole subpage function doesn't work.  That's not something IMO that should block this, but I think it's a definite bug to deal with at some point.

siddharthvp commented:
> My only other real nitpick is that if someone opens the dbatch menu, loads the subpage box, closes the dbatch menu, then re-opens it, the whole subpage function doesn't work.

Nice catch! Fixed. Will look at other comments when I have time.

siddharthvp commented on morebits.js:
```diff
@@ -3772,18 +3772,17 @@ Morebits.checkboxShiftClickSupport = function (jQuerySelector, jQueryContext) {
 
 				for (i = start; i <= finish; i++) {
-					cbs[i].checked = endState;
+					cbs[i].click();
```
done; earlier i failed to consider the use of shift-click over a range of checkboxes that are not all checked or all unchecked.

siddharthvp commented on morebits.js:
```diff
@@ -3772,18 +3772,17 @@ Morebits.checkboxShiftClickSupport = function (jQuerySelector, jQueryContext) {
 				if (index < lastIndex) {
 					start = index + 1;
-					finish = lastIndex;
+					finish = lastIndex - 1;
```
With your suggestion above of checking state before clicking, I think this `-1 / `+1` fix is no longer required.

Amorymeltzer replied:
@siddharthvp 	Did you mean to change the ±1 in the `else`?https://github.com/azatoth/twinkle/pull/581/files#diff-371d1913aa0a4d389d417d9bfda0e779R3781

siddharthvp replied:
looks like i screwed up a bit while changing it back to way it was originally. fixed it now.

Amorymeltzer replied:
Thanks, looks good now

siddharthvp commented:
> Will look at other comments when I have time.

all resolved.

Amorymeltzer commented:
💯 work on this
